builtin/provisioners/chef/resource_provisioner.go

```diff
@@ -0,0 +1,413 @@
+		p.installChefClient = p.sshInstallChefClient
+		p.createConfigFiles = p.sshCreateConfigFiles
+		p.runChefClient = p.runChefClientFunc(linuxConfDir)
+	case "winrm":
+		p.PreventSudo = true
+		p.installChefClient = p.winrmInstallChefClient
+		p.createConfigFiles = p.winrmCreateConfigFiles
+		p.runChefClient = p.runChefClientFunc(windowsConfDir)
+	default:
+		return fmt.Errorf("Unsupported connection type: %s", s.Ephemeral.ConnInfo["type"])
+	}
+
+	// Get a new communicator
+	comm, err := communicator.New(s)
+	if err != nil {
+		return err
+	}
+
+	// Wait and retry until we establish the connection
+	err = retryFunc(comm.Timeout(), func() error {
+		err := comm.Connect(o)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	defer comm.Disconnect()
+
+	if !p.SkipInstall {
+		if err := p.installChefClient(o, comm); err != nil {
+			return err
+		}
+	}
+
+	o.Output("Creating configuration files...")
+	if err := p.createConfigFiles(o, comm); err != nil {
+		return err
+	}
+
+	o.Output("Starting initial Chef-Client run...")
+	if err := p.runChefClient(o, comm); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Validate checks if the required arguments are configured
+func (r *ResourceProvisioner) Validate(c *terraform.ResourceConfig) (ws []string, es []error) {
+	p, err := r.decodeConfig(c)
+	if err != nil {
+		es = append(es, err)
+		return ws, es
+	}
+
+	if p.NodeName == "" {
+		es = append(es, fmt.Errorf("Key not found: node_name"))
+	}
+	if p.RunList == nil {
+		es = append(es, fmt.Errorf("Key not found: run_list"))
+	}
+	if p.ServerURL == "" {
+		es = append(es, fmt.Errorf("Key not found: server_url"))
+	}
+	if p.ValidationClientName == "" {
+		es = append(es, fmt.Errorf("Key not found: validation_client_name"))
+	}
+	if p.ValidationKeyPath == "" {
+		es = append(es, fmt.Errorf("Key not found: validation_key_path"))
+	}
+
+	return ws, es
+}
+
+func (r *ResourceProvisioner) decodeConfig(c *terraform.ResourceConfig) (*Provisioner, error) {
+	p := new(Provisioner)
+
+	decConf := &mapstructure.DecoderConfig{
+		ErrorUnused:      true,
+		WeaklyTypedInput: true,
+		Result:           p,
+	}
+	dec, err := mapstructure.NewDecoder(decConf)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := dec.Decode(c.Raw); err != nil {
+		return nil, err
+	}
+
+	if p.Environment == "" {
+		p.Environment = defaultEnv
+	}
+
+	if attrs, ok := c.Raw["attributes"]; ok {
+		p.Attributes, err = rawToJSON(attrs)
+		if err != nil {
+			return nil, fmt.Errorf("Error parsing the attributes: %v", err)
+		}
+	}
+
+	return p, nil
+}
+
+func rawToJSON(raw interface{}) (interface{}, error) {
+	switch s := raw.(type) {
+	case []map[string]interface{}:
+		if len(s) != 1 {
+			return nil, errors.New("unexpected input while parsing raw config to JSON")
+		}
+
+		var err error
+		for k, v := range s[0] {
+			s[0][k], err = rawToJSON(v)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return s[0], nil
+	default:
+		return raw, nil
+	}
+}
```

Comment:
Maybe a look at the tests will help giving some insights. Take for example this [test](https://github.com/svanharmelen/terraform/blob/f-chef-client-provisioner/builtin/provisioners/chef/ssh_provisioner_test.go#L200). If you look at what is going into the test you'll notice it's a bunch of `[]map[string]interface{}`'es which are generated by TF reading in your config file which started out like something like this:

```go
    provisioner "chef"  {
        attributes {
          "key1" {
            "subkey1" {
              "subkey2a" = ["val1", "val2", "val3"],
                "subkey2b" {
                  "subkey3" = "value3"
                }
              }
           }
        }
        node_name = "webserver1"
        server_url = "https://chef.company.com/organizations/org1"
        validation_client_name = "chef-validator"
        validation_key_path = "../chef-validator.pem"
    }
```

And then ended up in TF like something like this:

```go
"attributes": []map[string]interface{}{
  map[string]interface{}{
    "key1": []map[string]interface{}{
      map[string]interface{}{
        "subkey1": []map[string]interface{}{
          map[string]interface{}{
            "subkey2a": []interface{}{
              "val1", "val2", "val3",
            },
            "subkey2b": []map[string]interface{}{
              map[string]interface{}{
                "subkey3": "value3",
              },
            },
          },
        },
      },
    },
    "key2": "value2",
  },
}
```

So as this part should be used as plain `JSON` in the `first-boot.json` file needed for the initial Chef run, this function converts the format above to the output seen [here](https://github.com/svanharmelen/terraform/blob/f-chef-client-provisioner/builtin/provisioners/chef/ssh_provisioner_test.go#L238):

```go
{
  "key1":{
    "subkey1":{
      "subkey2a":["val1", "val2", "val3"],
      "subkey2b":{
        "subkey3":"value3"
      }
    }
  },
  "key2":"value2"
}
```

So that's about the what and why. As for the how... You can see the input is very consistent (everything is a `[]map[string]interface{}`) so given how Go encodes/[decodes arbitrary data](http://blog.golang.org/json-and-go#TOC_5.) you can see we only have to convert the slices of maps to just maps.

So I check for the length to confirm there is exactly 1 item in the slice and then return that one item. And by also doing this recursively for any values inside that map we iterate through the whole map ending up with a value `map[string]interface{}` which can be properly mashalled into JSON.